New Artefact: Dreamshard Necklace

amcnicky opened the pull request:
Still work to be done on the relevant vaults before this would likely be acceptable quality-wise, but opening the PR for feedback at this point. I will preface by saying that this item may well be a bit 'out there' design-wise, or even theme-wise, to the extent that I of course won't be offended if it's an angle that just isn't compatible with mainline crawl. I did have a lot of fun putting this together though, and think that it has the potential to add a unique and fun experience for the player in the games where this item generates and is found.

Overall vision: escaping your enemies by generating an entrance to another plane of existence is a pretty cool concept. It's also rare (zig figurine) to the extent that many (most?) players (those who avoid Tomb and haven't completed a zig) haven't had chance to use an escape with this tactical profile. The dreamshard necklace aims to give the player a similar option for escape, alongside some loot, and a wild (but ultimately short) flavourful journey through the Realm of Dreams, a new portal branch that can only be entered by breaking this item (one-time use).

Flavour vision: the dreamshard necklace is a tiny tear in the fabric of reality, crystalised within a fragile gemstone. Wearing the necklace regenerates both physical and magical reserves, as the wild auras of the realm of dreams passively imbue the wearer. Shattering the necklace destroys it forever, and unleashes the tear in reality, which grows to a portal just large enough to fit through. Entering the portal transports the player to the realm of dreams, thus triggering a series of short experiences that combine novel transformations, monsters, loot, and flavour.

Gameplay vision: primarily, this item gives the player a 2-turn escape: drop the portal, then step through it. The realm of dreams has dangers, but as long as the player keeps moving, they should be able to avoid damage. There is at least one dream in which the player will be able to rest back to full. This makes the item very strong, and though balance may be needed in this area, there are a few trade-offs.
1. This is a draining necklace, and so is difficult to swap on and off for fights where the player thinks it may be needed (with the goal of reducing tedium).
2. HP & MP regen, whilst nice enough that the player should be happy wearing it for a portion of the game, isn't end-game quality. There are no resistances on the item either. Combined, therefore, 1&2 mean that to have this escape at the player's disposal, they must commit to wearing the necklace, and having it take up what may well be a slot for which other meaningful choices and alternatives are available. As such, I think whilst it may be 'no brainer' to wear this artefact early (true of many good amulets when found early), the decision will not remain a 'no brainer' as alternatives with better resistances and base types are discovered. It poses the strategic question of 'would you rather have a one time use full heal taking 2 turns, or would you rather wear an amulet with good resistances/base type so that you're less likely to put yourself in a position of needing that heal in the first place?'

The other aspect to the item is the realm of dreams itself. The design goals for this 5-(short)-floor portal are to:
 - eventually give the player an opportunity to recover health and mana
 - whilst not making it trivial to do so on initial floors
 - whilst giving the opportunity to choose some loot as the player progresses through the series of 'dreams'
 - all wrapped up in an experience that feels unique, and befitting a rare item that the player will only have the opportunity to interact with quite infrequently. i.e. making it memorable.

The 'dreams' so far are:
 - You are a butterfly chased by ogres with butterfly nets
 - You are a pig chased around by a butcher whilst hungry diners watch
 - You are one of the floating blades in the Hall of Blades, floating aimlessly with no adventurer in sight
 - You are a brain worm, devouring some delicious brain
 - You are an orb of fire, leaving your homeland in search of adventure

Hope you all enjoy the idea! Always happy to iterate based on feedback.

![urand_dreamshard](https://user-images.githubusercontent.com/69476547/130215585-1ea050c6-b4cf-40b8-8ae2-46da56884281.png)
![dreams_portal](https://user-images.githubusercontent.com/69476547/130215610-4954d7b7-0a47-4780-9687-f63dc538333e.png)
![ogre_catcher](https://user-images.githubusercontent.com/69476547/130215644-8c82eb13-f8a9-434e-97b6-b6ab6da91ea7.png)
![brain_worm_form](https://user-images.githubusercontent.com/69476547/130215737-129b664b-f5a6-4d17-8c0b-d7564c1e8064.png)
![orb_of_fire_form](https://user-images.githubusercontent.com/69476547/130215741-803db95b-7bee-48b1-b7b5-f09f596fbec7.png)
![spectral_axe_form](https://user-images.githubusercontent.com/69476547/130215749-fe48014d-103c-429e-b9a7-6d7a33975883.png)
![butterfly_form](https://user-images.githubusercontent.com/69476547/130215774-e906554b-4c82-48c7-9a91-58f1479f94d3.png)
![pig_form](https://user-images.githubusercontent.com/69476547/130215782-7d2eca05-7904-4de7-976c-b02265dc752f.png)

nojaa3 reviewed (changes requested):
I did quick review mainly consisting of problems of Dreams, mainly the transformations permanency from view of Transmuters or other form users. **No coding review was done and binary files weren't checked!** You will need check from a real developer, but you probably need scrap the dream-transformations first and find a substitute monster for Ogre catchers.

Tips: lot of later monsters can attack multiple times, letting them have permanent debuffs might nerf usually unusable monsters to acceptable ones. The instruction from that are found from vault making guide, I think.

nojaa3 commented:
I am not a developer, but in my view this is a horrible idea. 

Of Amulets two-turn escape property is strong enough and regeneration amulets are desirable even without additional resistances. It doesn't need any more properties. Rest of the comments were put in review.

ebering commented on crawl-ref/source/dat/descript/monsters.txt:
```diff
@@ -1785,6 +1785,11 @@ ogre
 tremendously strong.
 %%%%
+ogre catcher
+
```
New monsters are absolutely "allowed". Save compatibility must be handled in the merge and is something to be aware of, not an obstacle to adding a new monster.

ebering commented on crawl-ref/source/dungeon.cc:
```diff
@@ -1970,7 +1970,7 @@ static bool _fixup_stone_stairs(bool preserve_vault_stairs,
         // Pan abuses stair placement for transits, as we want connectivity
         // checks.
-        needed_stairs = you.depth == 1
+        needed_stairs = (you.depth == 1 || player_in_branch(BRANCH_DREAMS))
```
This comment doesn't make sense in the context of what `_fixup_stone_stairs` does for the level builder and how this PR uses the level builder.

ebering commented on crawl-ref/source/art-func.h:
```diff
@@ -1620,3 +1622,47 @@ static void _SEVEN_LEAGUE_BOOTS_equip(item_def */*item*/, bool *show_msgs,
+{
+    if (_evoke_dreamshard_necklace())
+    {
+        ASSERT(in_inventory(*item));
```
`ASSERTS` are used to guard against impossible situations where the game *should* crash. In this case this ASSERT is guarding against the possibility of using the dreamshard necklace while it is not in player inventory.

ebering commented:
@nojaa3 thank you for your interest in the crawl project and your community engagement. However, please don't presume to speak for the dev team, speculations that require a "real developer"'s input are best left unsaid. In this case, adding a new monster or branch is a fine thing to do and while it does have save compatibility considerations that's something that can be dealt with.

Additionally, when giving feedback remember to keep it constructive and positive. "This is horrible" can come off as rude or an insult, compared to "This is a bad fit for mainline crawl" or "It seems like any version of this design is impossibly overpowered".

Finally, its important to separate out "I wouldn't use this on a transmuter"-type gameplay theory crafting from design feedback. The fact that some items are better or worse for certain characters is usually seen as a positive in crawl development: we don't like items that are so good you should always swap to them.

ebering commented:
@amcnicky When answering your questions about dlua and how to modify the player with vault code I guessed you were going down a pretty "out there" design road. Indeed, this is a pretty wild idea! Currently too enamored of the flavour and cool factor to think straight about gameplay.

I'm about to playtest a battle against Milo, so I'm a bit behind in review and feedback. If you want some high level input before you put in more effort stop by #crawl-dev . The team is active enough these days that you should find someone available for a quick spitball regardless of timezone.

PleasingFungus commented:
Wow, this is a cool idea! I'd need to do a deeper dive to give really good opinions, but some quick thoughts:

> This is a draining necklace, and so is difficult to swap on and off for fights where the player thinks it may be needed (with the goal of reducing tedium).

Good instinct, but now that amulets take 5 + 5 turns to swap, I don't think this should be needed.

> HP & MP regen, whilst nice enough that the player should be happy wearing it for a portion of the game, isn't end-game quality. 

I think this could be toned down further. A strong escape ability is something you want a strong tradeoff for, and most of the game isn't "end-game" :P 

> The 'dreams' so far are...

these themes rule :)

PleasingFungus commented:
Spicy variant idea: make this a Fragile ring which, when removed, generates the portal. (Not sure this is a *better* design, but if you're really worried about swaps..?)

amcnicky commented on crawl-ref/source/dat/des/portals/dreams.des:
```diff
@@ -0,0 +1,231 @@
+# 1. to give the player a somewhat safe on-demand (but one-time use) portal
+#    that they can step into as a very strong 2-turn escape like a ziggurine.
+#    (though it is harder to rest on the spot than in zig:1)
+# 2. to provide a good level of loot (less than a trove, more than the
```
It's a good question - what does the loot add to the item? I think it gives the player something to 'play' in the first few dreams - i.e. what can they collect without getting netted/butchered? Thereafter, the loot arguably adds less to the experience, although I do quite like the brain worm acquiring 'knowledge' (books) from within the brain dream. Perhaps the direction is to tone down the loot a little based on play-testing (open to balance suggestions on that!), but in lieu of a reason for why loot on this item makes the item worse for crawl in general, I think the presence of loot adds rather than detracts from the design.

amcnicky commented:
> Wow, this is a cool idea! I'd need to do a deeper dive to give really good opinions, but some quick thoughts:
> 
> > This is a draining necklace, and so is difficult to swap on and off for fights where the player thinks it may be needed (with the goal of reducing tedium).
> 
> Good instinct, but now that amulets take 5 + 5 turns to swap, I don't think this should be needed.
> 
> > HP & MP regen, whilst nice enough that the player should be happy wearing it for a portion of the game, isn't end-game quality.
> 
> I think this could be toned down further. A strong escape ability is something you want a strong tradeoff for, and most of the game isn't "end-game" :P
> 
> > The 'dreams' so far are...
> 
> these themes rule :)

Thanks for the feedback (to everyone who took the time to give their thoughts)! I hadn't spotted the amulet swap change to 5 turns, you're right that does go a long way to reducing the incentive for tedium. Potentially might be worth keeping the draining though to prevent the admittedly lesser tedium of seeing a dangerous fight then running away to reset and come back with the amulet on (actually thinking about this now, does anyone actually do this? probably not). Feels like in the interest of streamlining the idea the player should be thinking about whether to wear the amulet from a strategic point of view rather than a tactical/per-combat point of view, since the artefact is already quite complicated, so anything to reduce the amount of 'other decisions' relating to it would be a good thing? Making it fragile as you say would be another option that makes sense to me - indeed the item was originally fragile in the initial design, but I tried to stay away from confusing the player between shattering it (evoke) and shattering it (taking it off - hey where's my portal?).

Gopall commented:
One tiny thing I noticed about the new forms. Butterflies are SIZE_TINY and not SIZE_SMALL.

Gopall commented:
The rework is also interesting, but aside from the connected theme it's basically a completely different item now.

amcnicky commented:
> The rework is also interesting, but aside from the connected theme it's basically a completely different item now.

I actually quite disagree (about the item being significantly different), though I do agree that it's hopefully interesting :)

Let me explain my thoughts on that and perhaps this will also act as a short briefing on the re-work for would-be reviewers of this new streamlined implementation.

The original gameplay vision for the item was:
>...this item gives the player a 2-turn escape: drop the portal, then step through it. The realm of dreams has dangers, but 
>as long as the player keeps moving, they should be able to avoid damage. There is at least one dream in which the player will >be able to rest back to full. This makes the item very strong, and though balance may be needed in this area, there are a few >trade-offs.
i.e. a 2-turn combination of actions that would eventually heal the player to full hp and mp (resting) before returning to the fight they originally fled from.

The reworked version of the item is very similar tactically in that case
- it heals the player between 50% and 100% of their hp in 1 turn (arguably stronger than the original design)
- it does not restore mp (weaker than the original design)
- it does not provide any loot to the player (again weaker)
- it no longer provides hp regen or mp regen, and in fact the base type does nothing (weaker)
- it still is destroyed after 1 use
- it still requires the strategic tradeoff between wearing an amulet that does nothing 99% of the time, and then saves you once, versus wearing a much better amulet such that you might not need saving.
- it still requires the player to actually use it (no auto-heal), we would see deaths from lcs on characters wearing this item etc.

And since a lot of the flavour (portal, dreams etc) is cut in this rework, a modest number of dream-themed allies are also summoned in the now, which provides (in my opinion) a great place for otherwise unused art to be added to the game.

So from a tactical perspective, it takes effect 1 turn quicker and now summons allies, but in compensation there is no loot, hp regen, mp regen, or mp heal in the main effect. Since the core effect 'dream themed strong heal, with dream flavour effects' doesnt really change that much, I therefore believe that this streamlined version would be more compatible with mainline crawl whilst still retaining interesting strategic choices for the player.

Item is ready for a second look by any devs who have time!

semitonal commented:
This definitely seems like a much simpler and more appropriate design for a lifesaving-style unrand effect, which I do think is a fun concept generally. Some minor notes on the rework - as mentioned in the todo comments, it'd need to have the heal disabled under DDoor, so prompting might be good there (as-is, you'd get healed and then DDoor's ongoing effect would reset your health immediately). The allies could also perhaps be suppressed under Sacrifice Love/Okawaru worship, to avoid being dropping hostiles on those characters.

The evocation is probably better implemented as an ability on the normal ability menu - I think the unrand evoke_func framework has now been fully removed (see e74157ce7a5) so it won't work currently.

Gopall commented:
Just a thought that came to mind. Would the removed dream levels be interesting (or suitable) as a portal branch?

amcnicky commented:
> This definitely seems like a much simpler and more appropriate design for a lifesaving-style unrand effect, which I do think is a fun concept generally. Some minor notes on the rework - as mentioned in the todo comments, it'd need to have the heal disabled under DDoor, so prompting might be good there (as-is, you'd get healed and then DDoor's ongoing effect would reset your health immediately). The allies could also perhaps be suppressed under Sacrifice Love/Okawaru worship, to avoid being dropping hostiles on those characters.
> 
> The evocation is probably better implemented as an ability on the normal ability menu - I think the unrand evoke_func framework has now been fully removed (see [e74157c](https://github.com/crawl/crawl/commit/e74157ce7a5cd120dc30e5e6355988f9f89bfcc2)) so it won't work currently.

Thanks for the feedback! I've now implemented DD compliance and an allies forbidden check.
Secondly, I've uncoupled the item from evoking, but also moved it away from being player-activated entirely, since when I started looking through how to link items to the abilities menu, it appeared a few similar items had been moved to a *effect instead (e.g. ratskin cloak). As such, the current design triggers the item on a lethal or near-lethal instance of damage instead of being player-activated. This means that the player no longer needs to remember to use it, which may or may not be a good thing, but at the very least streamlines the item further.

I've also added a colourful visual effect to go along with the activation, fitting the description of 'a torrent of colour' - no gameplay impact.

amcnicky commented:
> Just a thought that came to mind. Would the removed dream levels be interesting (or suitable) as a portal branch?

Since my original dream levels were not intended to be actual combat vaults, anyone implementing that portal branch would be inheriting only the theme from this PR really. The original levels probably wouldn't make the best actual combat gameplay, but I would be thrilled if someone were to take the realm of dreams theme and make a portal with it! The portal tile is already done ;)

amcnicky commented:
This PR is ready for another look when any of the core team have time :)

semitonal commented:
I'll try and take a closer look at this soon, but some brief discussion in #crawl-dev raised the point that the allies it now summons might be a bit much - they're very tough with high damage, and they also don't scale depending on depth or XL. PleasingFungus suggested that they might be more appropriate as Shadow Creatures (which fit the theme and would make them automatically scale to the current depth), rather than being a fixed new monster type.

amcnicky commented:
> I'll try and take a closer look at this soon, but some brief discussion in #crawl-dev raised the point that the allies it now summons might be a bit much - they're very tough with high damage, and they also don't scale depending on depth or XL. PleasingFungus suggested that they might be more appropriate as Shadow Creatures (which fit the theme and would make them automatically scale to the current depth), rather than being a fixed new monster type.

Cool! That makes sense on the tuning of allies. Shame to see some of the unused art go, but the suggestion re: shadow creatures is a good one and probably less confusing for the player overall. I've updated the spawns to follow that suggestion, tidied some legacy code no longer needed, and adjusted descriptions/print statements to keep the item's theme cohesive in light of the move to shadow creatures.

Sorry for the late reply - work gets quite busy this time of year!

ebering commented:
I've picked this up to try to finish it. The visuals, while neat, are pretty extensive and I think overkill. There's still some cleanup of old code bouncing around but I can take care of it, as well as simplifying the checks to apply it.

I think what Kate was suggesting with Death's Door is that casting DDoor while wearing the necklace should prompt reminding the user that their necklace won't work while under DDoor. I'll implement the check that way.

ebering commented:
Merged with some tweaks, simplified visuals, additional checks and some code refactoring. Thanks for the idea!
